Multi sonar support

andresR8 opened the pull request:
Add the specific ID for each I2C sensor connected to the bus.

- Allows receives individuall data of multiple i2c maxbotics sensors over mavlink distance sensor message.
- Allows  split the individuall measurements in mavros with px4_config.yaml :

![screenshot from 2017-07-31 14-03-06](https://user-images.githubusercontent.com/17345869/28798602-d7d02e80-75f9-11e7-9a05-8fb281643be4.png)

dagar commented:
Why did you create a new px4fmu-v2 config? If you ran out of flash try rebasing on master. sdlog2 was removed from the default config.

andresR8 commented:
I think that would be easier for devs, who need the sonar support, compile the code using this new config file instead of edit the default config file. However this file is optional.

TSC21 commented:
I like the idea but I support that the ID should be based on the publish order to the uORB stream. So the first driver to publish to a uORB topic shouls get the ID 0, while the others should verify if there is a uORB distance sensor msg, and get an ID +1 higher than the distance sensor ID's being published. This gives consistancy to the framework and the sonars are not the only sensors that can be connected to the FCU (you can have a LIDAR connected also on the i2c bus but as it is in another driver, this same strategy won't update its ID).

TSC21 commented:
> I think that would be easier for devs, who need the sonar support, compile the code using this new config file instead of edit the default config file. However this file is optional.

If you want to suggest another config, do that on a seperate PR.

andresR8 commented:
In this sense, we could use the ID as the I2C sensor address that should be unique for each device on the bus.

 I'm going to create another PR for the new configuration file.

dagar commented:
What do you mean "instead of edit the default config file"? If you rebase on master you should have enough space to include this in the default config.

andresR8 commented:
@dagar What I mean is that if anyone will need a sonar support they just have to type make px4fmu-v2_mb12xx.  However  I've removed the new config file for this PR.  Also I've replaced the sensor id to use the I2C addess.

TSC21 commented:
I still think that using the I2C address as the sensor ID is not the best approach, as you also have distance sensors that can be connected to UART or the ADC and they are not mapped the same way. As I said, the way to go should be adding a probe for distance sensor uORB topic publishers that in a first opportunity verifies if one exists, if not creates advertises to it with ID 0, if yes just increments the ID relative to the highest ID detected beign advertised.

TSC21 commented:
Also that ID should be kept unique after FCU restart. The way to do it in that case should be associate the I2C address with the ID and kept that info on ROM.

andresR8 commented:
The thing is that in this moment when I connect multiple I2C sensors, only one uORB topic is published as distance_sensor0. So the individual data cannot be differentiated

TSC21 commented:
In the case of the same driver for mutiple sensors, you can store the ID/address association and advertise them. You can choose in that case if it is in a first-in-first-served based or based on the value of the address (lowest value of the address gets lowest value ID). Still, you will need to add the verification for another distance sensor being advertised and also the other drivers should be aware of the ones being advertised by this driver. So, as is, this PR is incomplete.

andresR8 commented:
@TSC21  You're right I'm going to try in that way

mhkabir commented:
I would propose to get rid of usage of the uORB instance ID as "device ID" for rangefinders. It is prone to reordering based on boot timing, and is not deterministic. We have a much better solution (see all the IMU drivers) which creates an unique device ID based on device type, internal/external bus, i2c address, etc. This will allow us to uniquely identify each sensor. It also does not change over reboots.

I don't have time to implement this in the rangefinder drivers, but if someone is willing to do it, here is how it's done : https://github.com/PX4/Firmware/blob/master/src/drivers/mpu9250/mpu9250.cpp#L170-L174

LorenzMeier commented:
@andresR8 Does this PR work for you now or not?

andresR8 commented:
@LorenzMeier  Yes, it's working for me, I'm building a rover with multiple(5) maxbotics's sonar sensors to avoid obstacles.

TSC21 commented:
@mhkabir I like your suggestion, but my main concern is related to rangefinders not sharing the I2C bus, like serial or ADC, or even those that are subscribed coming from the Mavlink stream. Shouldn't we have a common way of identifying all the distance sensors in a consistant way?

andresR8 commented:
@TSC21 what if the other sensors like ADC, etc. starts the id numeration from >= 0x71

TSC21 commented:
You will have to find a way of automating the atributtion of the ID's. The same applies for the ones coming from the Mavlink stream.

mhkabir commented:
> @mhkabir I like your suggestion, but my main concern is related to rangefinders not sharing the I2C bus, like serial or ADC, or even those that are subscribed coming from the Mavlink stream. Shouldn't we have a common way of identifying all the distance sensors in a consistant way?

@TSC21 My proposal deals with exactly that. Serial/ADC can just be different "bus types", and as long as you can somehow identify each sensor individually at the interface level (for e.g Mavlink, this can be the current ID field we have), you can use that identifier to construct the unique ID.

TSC21 commented:
@mhkabir yeah I saw that there's a `bus_type` field on the `devid_s` structure. So yeah I believe that's also the best option and the most consistent. @andresR8 maybe you can check https://github.com/PX4/Firmware/blob/master/src/drivers/device/vdev.h and do something similar for rangefinders in general?

TSC21 commented:
@andresR8 did you make any progress on this?

TSC21 commented:
@andresR8 do you plan to update this PR?
